### PR TITLE
Added notes for wk3

### PR DESCRIPTION
**Pull request by stevendes**
Still fixing some issues with Valgrind

### CODE REVIEW

**Comment by stevendes** [comment]
Added the Valgrind part. Ready for review

File: milner_desvars/wk3-notes.md [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+
+rosrun is a command from the rosbash package that allows you to run a single node from a package without needing to give it's full path or cd/roscd there first.
+
+Usage 
+
+`rosrun <package> <executable>`
+
+Ex.
+
+`rosrun gazebo_ros gazebo`
```
Comment (eborghi10) [inline milner_desvars/wk3-notes.md]: Could you use an example from the project?

File: milner_desvars/wk3-notes.md [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+
+`roslaunch ca_gazebo create_house.launch`
+
+## Difference between rosrun and roslaunch commands
+
+The main difference is that rosrun can only run one node at the time, and also need the master to be initialized (with `roscore`).
+
+roslaunch, on the other hand, takes an XML file that has all the information about what nodes to launch, so it can launch multiples nodes at once and also automatically runs the master if it is not already running.
+
+## ROS Architecture
```
Comment (eborghi10) [inline milner_desvars/wk3-notes.md]: What is this? Why did you add it?

File: milner_desvars/wk3-notes.md [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+
+### turtlebot
+
+In the project turtlebot, also the launch files are organized in the respective packages, the difference resides in that, if a launch file need to include another one, the included launch file  is on a folder called "include" in the same folder as the original launch file, also the format of the file is `.launch.xml`, different from the `.launch` original.
+
+## Profiling and debugging tools?
+
+### Valgrind
+
+The Valgrind tool suite provides several debugging and profiling tools that help you make your programs faster and more correct.
```
Comment (eborghi10) [inline milner_desvars/wk3-notes.md]: I don't like this definition because it doesn't reflect its usage. Check [this one](https://en.wikipedia.org/wiki/Valgrind).

File: milner_desvars/wk3-notes.md [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+
+#### Profiling a node with valgrind while launching ROS
+
+We can attach the diagnosis tool to the node in the launch file adding the following line to the launch file:
+
+`launch-prefix="valgrind"`
```
Comment (eborghi10) [inline milner_desvars/wk3-notes.md:62]: What else you can do? Any other argument?

File: milner_desvars/wk3-notes.md [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+
+Ex.
+
+```
+<node name="ca_driver" pkg="ca_driver" type="ca_driver" output="screen">
+    <rosparam command="load" file="$(arg config)" />
+    <param name="robot_model" value="CREATE_2" />
+    launch-prefix="valgrind"
```
Comment (eborghi10) [inline milner_desvars/wk3-notes.md]: Did you try this? This doesn't work.

Comment (eborghi10) [reply milner_desvars/wk3-notes.md]: Check the [documentation](http://wiki.ros.org/roslaunch/XML/node).

File: milner_desvars/wk3-notes.md [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+==30266==    still reachable: 118,198 bytes in 395 blocks
+==30266==         suppressed: 0 bytes in 0 blocks
+==30266== Rerun with --leak-check=full to see details of leaked memory
+==30266== 
+==30266== For counts of detected and suppressed errors, rerun with: -v
+==30266== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
+
+```
+
+The interpretation of the message we can do in the summary.
```
Comment (eborghi10) [inline milner_desvars/wk3-notes.md]: This is very hard to read/understand. Use `kcachegrind` to read its logs.

File: milner_desvars/wk3-notes.md [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+`launch-prefix="xterm -e gdb --args"`
+
+Ex.
+
+```
+<node name="ca_driver" pkg="ca_driver" type="ca_driver" output="screen">
+    <rosparam command="load" file="$(arg config)" />
+    <param name="robot_model" value="CREATE_2" />
+    launch-prefix="xterm -e gdb --args"
+  </node>
```
Comment (eborghi10) [inline milner_desvars/wk3-notes.md]: Code is not aligned.

File: milner_desvars/wk3-notes.md [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+```
+
+With this launch prefix, a new xterm terminal will be created with the node attached to gdb
```
Comment (eborghi10) [inline milner_desvars/wk3-notes.md]: Missing ending dot.

File: milner_desvars/wk3-notes.md [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+  <arg name="y"/>
+  <arg name="Y"/>
+
+  <arg name="ns"/>
+
+  <node pkg="ca_move_base" type="send_goal" name="navigation_goals" 
+        args="$(arg x) $(arg y) $(arg Y)" ns="$(arg ns)" launch-prefix="xterm -e gdb --args"/>
+</launch>
+```
+So now when we launch the node, it also would launch gdb on a xter, so we are going to do that, launching with arbitrary arguments for testing:
```
Comment (eborghi10) [inline milner_desvars/wk3-notes.md]: Typo: xterm

File: milner_desvars/wk3-notes.md [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+Other thing we can do is put a breakpoint, we do that using the `break` command, followed by the number of the line in which we are going to put the breakpoint:
+
+`break 14`
+
+Those are some of the basics tools to debug in pdb, you can also integrate pdb into the program code by editing the file, for more information and a table of utilities you can visit the following page:
+
+[https://www.digitalocean.com/community/tutorials/how-to-use-the-python-debugger](https://www.digitalocean.com/community/tutorials/how-to-use-the-python-debugger)
+
+
+![pdb](media/pdb.png)
```
Comment (eborghi10) [inline milner_desvars/wk3-notes.md:259]: this image has some issues.

File: milner_desvars/wk3-notes.md [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+
+### rqt
```
Comment (eborghi10) [inline milner_desvars/wk3-notes.md]: why is this here?

File: milner_desvars/wk3-notes.md [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+
+
+# Source
+
+
+[https://www.geeksforgeeks.org/gdb-step-by-step-introduction/](https://www.geeksforgeeks.org/gdb-step-by-step-introduction/)
+
+[http://wiki.ros.org/rqt](http://wiki.ros.org/rqt)
+
+[http://valgrind.org/docs/manual/quick-start.html]http://valgrind.org/docs/manual/quick-start.html)
```
Comment (eborghi10) [inline milner_desvars/wk3-notes.md]: Wrong format. Please check.

**Review by eborghi10 — CHANGES_REQUESTED** [review]
I made some comments, PTAL.

**Comment by eborghi10** [comment]
Could you put a picture with the launch structure?

File: milner_desvars/wk3-notes.md [hunk 13]
```diff
@@ -88,31 +51,46 @@ In the project turtlebot, also the launch files are organized in the respective
...
 Ex.
 
 ```
-<node name="ca_driver" pkg="ca_driver" type="ca_driver" output="screen">
-    <rosparam command="load" file="$(arg config)" />
-    <param name="robot_model" value="CREATE_2" />
-    launch-prefix="valgrind"
-  </node>
+  <node pkg="ca_move_base" type="send_goal" name="navigation_goals" 
+        args="$(arg x) $(arg y) $(arg Y)" ns="$(arg ns)" launch-prefix="valgrind/>
```
Comment (eborghi10) [inline milner_desvars/wk3-notes.md]: Missing `"`

File: milner_desvars/wk3-notes.md [hunk 14]
```diff
@@ -88,31 +51,46 @@ In the project turtlebot, also the launch files are organized in the respective
...
+</launch>
   ```
 
 ### Example of profiling with Valgrind
 
-First of all we need to install Valgrind, we are doing that with the following command:
+First of all, we need to install the tools that would help us to analyze this node, these are, valgrind to do the profiling, and kchegrind and graphviz to represent them in a graphical interphase:
```
Comment (eborghi10) [inline milner_desvars/wk3-notes.md]: Typo: Interface

Please, take some time to check the grammar. We already reviewed this word before.

File: milner_desvars/wk3-notes.md [hunk 15]
```diff
@@ -185,10 +156,10 @@ Ex.
     <rosparam command="load" file="$(arg config)" />
     <param name="robot_model" value="CREATE_2" />
     launch-prefix="xterm -e gdb --args"
```
Comment (eborghi10) [inline milner_desvars/wk3-notes.md]: As it is, it won't work. Please test it or read the documentation.

**Review by eborghi10 — CHANGES_REQUESTED** [review]
Some changes need to be done, PTAL.

File: milner_desvars/wk3-notes.md [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+
+So we set our new breakpoint by using `break 10`, and run the program again, once it stops in the breakpoint, we use the following command to see the value of `NODE_NAME`:
+
+`disp/s NODE_NAME`
+
+And we get the value `"navigation_goals"`, we set the format of the output by using the `/s` and obtain a string as an output, manipulating this value you can obtain the output in different formats such as int or hexadecimal.
+
+As we show this are the basic movements to start debugging a node in C++, for more commands to use in `gdb` you can see them here:
+
+[http://www.yolinux.com/TUTORIALS/GDB-Commands.html](http://www.yolinux.com/TUTORIALS/GDB-Commands.html)
```
Comment (eborghi10) [inline milner_desvars/wk3-notes.md]: Add the link to the word "here".

**Review by eborghi10 — CHANGES_REQUESTED** [review]
One more comment.

**Review by eborghi10 — APPROVED** [review]

**Comment by stevendes** [comment]
Added rqt_launch info

**Comment by eborghi10** [comment]
> Still fixing some issues with Valgrind

**The description of the PR must explain its purpose.**

File: milner_desvars/wk3-notes.md [hunk 17]
```diff
@@ -0,0 +1,294 @@
...
+
+rqt_launch is an rqt plugin that allows you to see and easy way to view nodes and also to run or end nodes individually.
+
+To run it, first you enter in rqt using on the console:
+
+`rqt` 
+
+After that, you select the `Launch(Experimental)` option in the Plugin tab.
+
+![rqtlaunch](media/rqtlaunch1.png)
```
Comment (eborghi10) [inline milner_desvars/wk3-notes.md:271]: The bottom half of the image is useless so check if you can remove it. A very small part of the image shows useful information.

**Review by eborghi10 — CHANGES_REQUESTED** [review]
One small comment, PTAL.

File: milner_desvars/wk3-notes.md (hunk 17 above)
Comment (stevendes) [reply milner_desvars/wk3-notes.md:271]: Addressed

Comment (eborghi10) [inline milner_desvars/wk3-notes.md:271]: Can you increase the size of the windows so the scroll bar disappears?

**Review by eborghi10 — CHANGES_REQUESTED** [review]

**Review by eborghi10 — APPROVED** [review]
